### PR TITLE
DeepCSV/DeepCMVA discriminators

### PR DESCRIPTION
**Pull request by mverzett**
Introduced DeepCSV discriminator taggers as simple arithmetic over the input probabilities.

The PR introduces three new bDiscriminators to the miniAOD event content, containing the DeepCSV discriminators computed starting from the DeepCSV probabilities. Only the miniAOD content is expected to change.

The change is implemented modifying the RECO sequences, but given the modules are never called they should not be scheduled.

### CODE REVIEW

**Comment by perrotta** [comment]
please test

**Comment by slava77** [comment]
On 10/17/17 5:21 AM, Mauro Verzetti wrote:
> Introduced DeepCSV/DeepCMVA discriminator taggers as simple arithmetic
> over the input probabilities
> 

Please help to improve self-documentation and add some details:
- a link to slides or a web-page describing the added discriminators
- incremental changes from this PR for available products or which new
products we should expect in standard miniAOD and RECO workflows

**Comment by perrotta** [comment]
@mverzett : the error is correlated to this PR, please have a look

**Comment by mverzett** [comment]
@perrotta, on lxplus I keep getting das errors, I checked to have a valid proxy. Is there something else I have to do in order to access the files?

**Comment by perrotta** [comment]
@mverzett  : it's also happening to me since yesterday, that's weird...

I solved it by explicitely adding the file names in the step2 config (e.g. --command "--filein /store/data/Run2016E/JetHT/AOD/18Apr2017-v1/100000/02FB8769-5433-E711-BBDE-0025905A6076.root" in the cmsDriver command for step2)

**Comment by mverzett** [comment]
@perrotta : even explicitly forcing the file name all I get is "FileOpenError" exceptions. I tryied different redirectors (cmseos, cms-xrd-global, cmsxrootd.fnal.gov) all with the same result. Any idea what's going on?

**Comment by mverzett** [comment]
@perrotta : I managed to run the tests on a different file and removed the problematic part, which was introduced by mistake

**Comment by perrotta** [comment]
please test
thanks @mverzett

File: RecoBTag/Combined/plugins/BuildFile.xml [hunk 1]
```diff
@@ -1,6 +1,7 @@
 <use name="FWCore/Framework"/>
 <use name="FWCore/MessageLogger"/>
 <library file="*.cc" name="RecoBTagCombinedPlugins">
+	<flags CXXFLAGS="-g"/>
```
Comment (slava77) [inline RecoBTag/Combined/plugins/BuildFile.xml]: should be removed

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
@mverzett please add the documentation requested in https://github.com/cms-sw/cmssw/pull/20940#issuecomment-337234889 to the description part of this PR

In particular, please explain what we should expect different in the in standard miniAOD and RECO workflows

**Comment by mverzett** [comment]
@perrotta sorry for the late reply, I changed the description accordingly.

File: RecoBTag/Combined/plugins/BTagProbabilityToDiscriminator.cc [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+	}
+}
+
+// ------------ method fills 'descriptions' with the allowed parameters for the module  ------------
+void
+BTagProbabilityToDiscriminator::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  //The following says we do not know what parameters are allowed so do no validation
+  // Please change this to state exactly what you do use, even if it is no parameters
+  edm::ParameterSetDescription desc;
+  desc.setUnknown();
```
Comment (perrotta) [inline RecoBTag/Combined/plugins/BTagProbabilityToDiscriminator.cc]: https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp#ParameterSets_that_Contain_Vecto

**Comment by perrotta** [comment]
Looking at the outputs of wf 10224 (TTbar PU) I can confirm that there are no additions to the reco outputs, while the miniAOD output shows the following increase  (340 bytes/evt):

-----------------------------------------------------------------
   or, B         new, B      delta, B   delta, %   deltaJ, %    branch 
-----------------------------------------------------------------
   1263.3 ->      1265.4          2      0.2   0.00     patJets_slimmedJetsAK8__RECO.
   3480.9 ->      3703.8        223      6.2   0.33     patJets_slimmedJets__RECO.
   2487.6 ->      2604.3        117      4.6   0.17     patJets_slimmedJetsPuppi__RECO.
-------------------------------------------------------------
    67991 ->       68331        340             0.5     ALL BRANCHES

On the other hand, it is not clear to me how the new discriminators are run in miniAOD and not in Reco. They are included in the pfDeepFlavourTask, which enters in btagging
(RecoBTag/​Configuration/​python/​RecoBTag_cff.py), which enters in its turn in the main reco sequence
(Configuration/StandardSequences/python/Reconstruction_cff.py)

Could you please point to me the switches in the configuration/code that allow the new discriminators to be produced in the miniAOD and not in reco?

**Comment by mverzett** [comment]
@perrotta , sorry for the long delay, I was involved in a school last week which took all my time. The discriminators do not show up in reco as they are not added in the [reco content](https://github.com/cms-sw/cmssw/blob/master/RecoBTag/Configuration/python/RecoBTag_EventContent_cff.py), hence the unscheduled mode should take care of ignoring them.
If you have a more sound/elegant solution to suggest I am more than happy to implement it.

**Comment by perrotta** [comment]
Thank you @mverzett 
Please also address the comment in  https://github.com/cms-sw/cmssw/pull/20940#pullrequestreview-71467570. i.e. please initialize the ParameterSetDescription with some default vector of PSets. Then I think we are done with this PR.

**Comment by perrotta** [comment]
@mverzett : what is the point of having the modules in the standard reco sequence if not needed/used?
Couldn't you just remove them, and they will only be put back if/when needed?

**Comment by mverzett** [comment]
@perrotta indeed that could work, but it would make the re-making of the discriminators a bit more messy, as the PAT JetTools (add/updateJetCollection) use the default sequences to look for the discriminator modules to be re-run

**Comment by slava77** [comment]
On 10/31/17 12:10 PM, Mauro Verzetti wrote:
> @perrotta <https://github.com/perrotta> indeed that could work, but it
> would make the re-making of the discriminators a bit more messy, as the
> PAT JetTools (add/updateJetCollection) use the default sequences to look
> for the discriminator modules to be re-run

Why can't these modules be added to the standard sequence and saved in
RECO/AOD?
Do they need inputs that do not exist in RECO?

File: RecoBTag/Combined/plugins/BTagProbabilityToDiscriminator.cc [hunk 3]
```diff
@@ -161,8 +161,54 @@ void
 BTagProbabilityToDiscriminator::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
   //The following says we do not know what parameters are allowed so do no validation
```
Comment (perrotta) [inline RecoBTag/Combined/plugins/BTagProbabilityToDiscriminator.cc]: These two comment lines are not valid any more and can now be removed

File: RecoBTag/Combined/plugins/BTagProbabilityToDiscriminator.cc [hunk 4]
```diff
@@ -161,8 +161,54 @@ void
...
   // Please change this to state exactly what you do use, even if it is no parameters
-  edm::ParameterSetDescription desc;
-  desc.setUnknown();
+	edm::ParameterSetDescription desc;
```
Comment (perrotta) [inline RecoBTag/Combined/plugins/BTagProbabilityToDiscriminator.cc]: Please indent coherently the code in the whole method: as it is now it is rather difficult to read it

**Comment by mverzett** [comment]
@slava77 They can be added in RECO/AOD, I was putting them in miniAOD only to save AOD space and have the chance to backport it to this year's miniAOD

**Comment by mverzett** [comment]
@perrotta I ran clang-format to get a decent and consistent indentation

**Comment by perrotta** [comment]
Thank you Mauro: it looks much better now!
Still, please remove commented out lines 180-183, or modify what s  
written there, since it not true any more now.

Mauro Verzetti <notifications@github.com> ha scritto:

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
"Reco comparison results: 30 differences found in the comparisons"

It looks like all pairDiscrivector's got emptied, if I understand correctly the output:

image

**Comment by mverzett** [comment]
@perrotta , what is the command to reproduce the problem?

**Comment by perrotta** [comment]
@mverzett : this happens in the miniAOD output of any workflow run in jenkins, see e.g.
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_10_0_X_2017-10-30-2300+20940/23380/validateJR/logRootQA.log
where you can read
JR results differ 2 all_mini_OldVSNew_Higgs200ChargedTauswf9p0
JR results differ 2 all_mini_OldVSNew_TTbarwf25p0
JR results differ 2 all_mini_OldVSNew_RunPhoton2012Bwf4p53
JR results differ 2 all_mini_OldVSNew_TTbarPUwf25202p0
JR results differ 2 all_mini_OldVSNew_TTbar13TeV2019wf11624p0
JR results differ 2 all_mini_OldVSNew_ZMM13TeV2017wf10042p0
JR results differ 2 all_mini_OldVSNew_TTbar14TeV2023D19wf20434p0
JR results differ 2 all_mini_OldVSNew_TTbar13TeVPU2017wf10224p0
JR results differ 2 all_mini_OldVSNew_ZMM13TeVwf1330p0
JR results differ 2 all_mini_OldVSNew_TTbar13TeV2018wf10824p0
JR results differ 2 all_mini_OldVSNew_TTbar14TeV2023D17wf20034p0
JR results differ 2 all_OldVSNew_RunJetHT2016EreMINIAODwf136p7611
JR results differ 2 all_mini_OldVSNew_TTbar13TeV2017wf10024p0
JR results differ 2 all_mini_OldVSNew_RunSinglePh2016Bwf136p731
JR results differ 2 all_mini_OldVSNew_RunSinglePh2017B136p788

You can check any of those workflows, then

**Comment by slava77** [comment]
> @slava77 They can be added in RECO/AOD, I was putting them in miniAOD only to save AOD space and have the chance to backport it to this year's miniAOD

well, if the module is not needed in AOD/RECO, it should be removed from the reco sequence.

**Comment by ferencek** [comment]
@mverzett, I'm not sure I understand why you need to keep these taggers in the reco sequence. All the taggers supported in PAT are defined in https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/PatAlgos/python/recoLayer0/bTagging_cff.py and the list is significantly longer than what we run in the standard reco. Yes, all the various cfi files need to be eventually imported in the main b-tagging cff file (https://github.com/cms-sw/cmssw/blob/master/RecoBTag/Configuration/python/RecoBTag_cff.py) to be accessible in PAT but there is no reason to put all the taggers in the standard sequences.

**Comment by perrotta** [comment]
@mverzett : about https://github.com/cms-sw/cmssw/pull/20940#issuecomment-340807008, Slava just made me notice that before https://github.com/cms-sw/cms-bot/pull/911 miniAOD plots were broken.

Therefore, what reported in  https://github.com/cms-sw/cmssw/pull/20940#issuecomment-340807008 is just the net effect of this PR, and not of your last commit only. Adding new discriminators the size of the vectors got probably out of scale (and not to 0...).

In summary, https://github.com/cms-sw/cmssw/pull/20940#issuecomment-340807008 doesn't looks like being a problem: sorry for the noise, and just cocentrate on the removal of the jet tageers from the reco sequence if not needed

**Comment by mverzett** [comment]
@perrotta, please disregard this comment, I was trying to go through without a rebase, but it looks like it is necessary. I will do it asap (most likely tomorrow morning)

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by mverzett** [comment]
@perrotta @slava77 As promised during the meeting I rebased to get rid of all the spurious additions.

**Comment by perrotta** [comment]
+1
- DeepCSV/DeepCMVA discriminators added to miniAOD as simple arithmetic over the input probabilities
- miniAOD content increases by 340 bytes (0.5%) in the 10224 (TTbar PU) wf
- With the last commit, those discriminators had been removed from the standard btag reco sequence as not needed there

File: RecoBTag/Combined/plugins/BTagProbabilityToDiscriminator.cc [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+  }
+}
+
+void BTagProbabilityToDiscriminator::produce(edm::Event &iEvent,
+                                             const edm::EventSetup &iSetup) {
+  std::unordered_map<std::string, edm::Handle<JetTagCollection>>
+      tags; // caches jet tags to avoid repetitions
+  size_t size = 0;
+  bool first = true;
+  for (auto entry : jet_tags_) {
```
Comment (davidlange6) [inline RecoBTag/Combined/plugins/BTagProbabilityToDiscriminator.cc]: this should be a const &
@mverzett

File: RecoBTag/Combined/plugins/BTagProbabilityToDiscriminator.cc [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+        *(tags.begin()->second)) // clone from the first element, will change
+                                 // the content later on
+                          );
+  }
+
+  // loop over jets
+  for (size_t idx = 0; idx < output_tags[0]->size(); idx++) {
+    auto key = output_tags[0]->key(idx); // index access should be OK as well,
+                                         // but a bit less safe, probably WAY
+                                         // faster
```
Comment (davidlange6) [inline RecoBTag/Combined/plugins/BTagProbabilityToDiscriminator.cc]: indeed, why not count on a sane interface from JetTagCollection?

File: RecoBTag/Combined/plugins/BTagProbabilityToDiscriminator.cc [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+
+    // loop over new discriminators to produce
+    for (size_t disc_idx = 0; disc_idx < output_tags.size(); disc_idx++) {
+      float numerator = 0;
+      for (auto &num : discrims_[disc_idx].numerator)
+        numerator += (*tags[num])[key];
+      float denominator = !discrims_[disc_idx].denominator.empty() ? 0 : 1;
+      for (auto &den : discrims_[disc_idx].denominator)
+        denominator += (*tags[den])[key];
+      float new_value = (denominator != 0 && numerator > 0) ? numerator / denominator : -10.;
```
Comment (davidlange6) [inline RecoBTag/Combined/plugins/BTagProbabilityToDiscriminator.cc]: I don't really follow the need to check that numerator>0? Is there an easy explanation?

File: RecoBTag/Combined/plugins/BTagProbabilityToDiscriminator.cc (hunk 6 above)
Comment (mverzett) [reply RecoBTag/Combined/plugins/BTagProbabilityToDiscriminator.cc]: @davidlange6 the index access assumes that all the jet tags passed follow the same sorting. Given that the probabilities should produced by the same module (now they are, cannot guarantee in future) I agree with you. I can change that if you want.

File: RecoBTag/Combined/plugins/BTagProbabilityToDiscriminator.cc (hunk 7 above)
Comment (mverzett) [reply RecoBTag/Combined/plugins/BTagProbabilityToDiscriminator.cc]: The probability < 0 is used to mark an undefined jet, where there is no information to make a decision (no tracks pass the track pre-selection and no secondary vertices are found). With this module the discriminator defaults to some number, sometimes positive (e.g. the charm tagger which is prob(c)/(prob(b)+prob(c) defaulting at 0.5), hence the need of that guard. I can add a comment.

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
+1
- Already signed in  https://github.com/cms-sw/cmssw/pull/20940#issuecomment-341877836
- All further changes applied after David's comments also look fine

**Comment by davidlange6** [comment]
merge
